prefect_kubernetes/utilities.py

```diff
@@ -180,3 +185,126 @@ def _slugify_label_value(value: str, max_length: int = 63) -> str:
     # Kubernetes to throw the validation error
 
     return slug
+
+
+class ResilientStreamWatcher:
+    """
+    A wrapper class around kuberenetes.watch.Watch that will reconnect on
+    certain exceptions.
+    """
+
+    DEFAULT_RECONNECT_EXCEPTIONS = (urllib3.exceptions.ProtocolError,)
+
+    def __init__(
+        self,
+        logger: Optional[logging.Logger] = None,
+        max_cache_size: int = 50000,
+        reconnect_exceptions: Optional[List[Type[Exception]]] = None,
+    ) -> None:
+        """
+        A utility class for managing streams of Kuberenetes API objects and logs
+
+        Attributes:
+            logger: A logger which will be used interally to log errors
+            max_cache_size: The maximum number of API objects to track in an
+                internal cache to help deduplicate results on stream reconnects
+            reconnect_exceptions: A list of exceptions that will cause the stream
+                to reconnect.
+        """
+
+        self.max_cache_size = max_cache_size
+        self.logger = logger
+        self.watch = watch.Watch()
+
+        reconnect_exceptions = (
+            reconnect_exceptions
+            if reconnect_exceptions is not None
+            else self.DEFAULT_RECONNECT_EXCEPTIONS
+        )
+        self.reconnect_exceptions = tuple(reconnect_exceptions)
+
+    def _stream(self, cache: Optional[FIFOCache], func: Callable, *args, **kwargs):
```

Comment:
Would this signature work? I'm thinking it'd be nice not to have to pass `None` for `cache` in cases it isn't needed.
```suggestion
    def _stream(self, func: Callable, *args, cache: Optional[FIFOCache] = None, **kwargs):
```